python/python/bystro/tada_mixture/mixture_tada_em.py

```diff
@@ -0,0 +1,394 @@
+
+        for i in myrange(td["n_iterations"]):
+            # E step compute most likely categories
+            Zprobs = np.zeros((N, K))
+            for k in range(K):
+                Lambda_k = Lambda_list_l[k]
+                alpha_k = 0.8 * sigmoid(alpha_list_l[k])  # Prob of being 0
+
+                log_likelihood_poisson = -1 * myloss(Lambda_k, data)
+                log_likelihood_poisson_w = log_likelihood_poisson + torch.log(
+                    1 - alpha_k
+                )
+
+                log_likelihood_0 = torch.log(
+                    alpha_k + (1 - alpha_k) * torch.exp(-1 * Lambda_k)
+                )
+                log_likelihood_0_b = torch.broadcast_to(
+                    log_likelihood_0, data.shape
+                )
+
+                log_likelihood_point = torch.where(
+                    data != 0, log_likelihood_poisson_w, log_likelihood_0_b
+                )
+
+                log_marginal = torch.sum(log_likelihood_point, axis=1)
+
+                z_probs_k = log_marginal + np.log(pi_[k])
+                Zprobs[:, k] = z_probs_k.detach().numpy()
+
+            Ez = np.argmax(Zprobs, axis=1)
+
+            # M step, maximize parameters
+            for k in range(K):
+                pi_[k] = np.mean(Ez == k)
+                data_sub = data[Ez == k]  # Selects relevant data
+
+                trainable_variables = [Lambda_list_l[k], alpha_list_l[k]]
+
+                optimizer = torch.optim.SGD(
+                    trainable_variables, lr=td["learning_rate"], momentum=0.8
+                )
+
+                for j in range(td["n_inner_iterations"]):
+                    Lambda_k = Lambda_list_l[k]
+                    alpha_k = 0.8 * sigmoid(alpha_list_l[k])  # Prob of being 0
+
+                    log_likelihood_poisson = -1 * myloss(Lambda_k, data_sub)
+                    log_likelihood_poisson_w = (
+                        log_likelihood_poisson + torch.log(1 - alpha_k)
+                    )
+
+                    log_likelihood_0 = torch.log(
+                        alpha_k + (1 - alpha_k) * torch.exp(-1 * Lambda_k)
+                    )
+                    log_likelihood_0_b = torch.broadcast_to(
+                        log_likelihood_0, data_sub.shape
+                    )
+
+                    log_likelihood_point = torch.where(
+                        data_sub != 0,
+                        log_likelihood_poisson_w,
+                        log_likelihood_0_b,
+                    )
+
+                    log_marginal = torch.sum(log_likelihood_point, axis=1)
+                    loss = -1 * torch.mean(log_marginal)
+
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+
+                    Lambda_k.requires_grad_(False)
+                    Lambda_k[Lambda_k < 0] = 0
+                    Lambda_k.requires_grad_(True)
+
+        self.pi = pi_
+        self.Lambda = -1000 * np.ones((K, p))
```

Comment:
why -1000?

Reply:
Make it obvious if there's a problem. Not feasible value.